File: overlay/python/hydra-kernel/src/main/python/hydra/lib/pairs.py
"""Python implementations of hydra.lib.pairs primitives."""

from __future__ import annotations
from collections.abc import Callable
from typing import TypeVar

A = TypeVar('A')
B = TypeVar('B')
C = TypeVar('C')
D = TypeVar('D')


def bimap(f: Callable[[A], C], g: Callable[[B], D], pair: tuple[A, B]) -> tuple[C, D]:
    """Map over both elements of a pair."""
    return (f(pair[0]), g(pair[1]))


def first(pair: tuple[A, B]) -> A:
    """Get the first element of a pair."""
    return pair[0]


def second(pair: tuple[A, B]) -> B:
    """Get the second element of a pair."""
    return pair[1]

File: overlay/typescript/hydra-kernel/src/main/typescript/hydra/lib/pairs.ts
// Hand-written runtime: hydra.lib.pairs primitives.
//
// `first` and `second` accept `unknown` (so kernel-routed `unknown`
// pairs type-check) and return `any` (so callers can use the result
// without TS18046 "X is of type unknown"). They're total at runtime for
// any 2-element array input.
export const first = (p: unknown): any => (p as readonly [any, any])[0];
export const second = (p: unknown): any => (p as readonly [any, any])[1];

export const swap = <A, B>(p: readonly [A, B]): readonly [B, A] => [p[1], p[0]];

export const bimap = <A, B, A2, B2>(f: (a: A) => A2, g: (b: B) => B2, p: readonly [A, B]): readonly [A2, B2] =>
  [f(p[0]), g(p[1])];
